src/fairseq2/datasets/preference.py

```diff
@@ -201,66 +204,142 @@ def create_reader(
+                    target_mask_rejected = torch.full([len(indices_rejected)], True)
+
+                total_tokens = (
+                    2 * len(source_indices)
+                    + len(target_indices_chosen)
+                    + len(target_indices_rejected)
+                )
 
-            # below is an example of using extras field of data reader options
-            if "keep_jsonl_keys" in options.extras:
-                jsonl_keys = options.extras["keep_jsonl_keys"]
-                if not (
-                    isinstance(jsonl_keys, list)
-                    and all(isinstance(i, str) for i in jsonl_keys)
-                ):
-                    raise ValueError(f"{jsonl_keys} must be a list of strings")
-                jsonl_content = {k: example.get(k, None) for k in jsonl_keys}
-            else:
-                jsonl_content = None
-
-            return {
-                "id": id_,
-                "indices_prompt": source_indices,
-                "indices_chosen": indices_chosen,
-                "indices_rejected": indices_rejected,
-                "reference_score_chosen": example.get("reference_score_chosen", None),
-                "reference_score_rejected": example.get(
-                    "reference_score_rejected", None
-                ),
-                "target_mask_chosen": target_mask_chosen,
-                "target_mask_rejected": target_mask_rejected,
-                "total_tokens": total_tokens,
-                "keep_jsonl_keys": jsonl_content,
-            }
-
-        builder.map(cat_source_and_target)
+                # below is an example of using extras field of data reader options
+                if "keep_jsonl_keys" in options.extras:
+                    jsonl_keys = options.extras["keep_jsonl_keys"]
+                    if not (
+                        isinstance(jsonl_keys, list)
+                        and all(isinstance(i, str) for i in jsonl_keys)
+                    ):
+                        raise ValueError(f"{jsonl_keys} must be a list of strings")
+                    jsonl_content = {k: example.get(k, None) for k in jsonl_keys}
+                else:
+                    jsonl_content = None
+
+                return {
+                    "id": id_,
+                    "indices_prompt": source_indices,
+                    "indices_chosen": indices_chosen,
+                    "indices_rejected": indices_rejected,
+                    "reference_score_chosen": example.get(
+                        "reference_score_chosen", None
+                    ),
+                    "reference_score_rejected": example.get(
+                        "reference_score_rejected", None
+                    ),
+                    "target_mask_chosen": target_mask_chosen,
+                    "target_mask_rejected": target_mask_rejected,
+                    "total_tokens": total_tokens,
+                    "keep_jsonl_keys": jsonl_content,
```

Comment:
these keys are also not returned in instruction.py. not sure if we want those

Reply:
yep we use total tokens in some of the metrics in preference recipe, and keep_jsonl_keys is used when we pass reference scores precomputed offline